Migrate the tests from etap to eunit

eiri opened the pull request:
All the etap tests converted to eunit.

This work is a part of COUCHDB-2590.

eiri committed:
Migrate the tests from etap to eunit

All the etap tests converted to eunit.

This work is a part of COUCHDB-2590.

kxepal commented on test/khash_test.erl:
```diff
@@ -0,0 +1,443 @@
+dict_fetch(_, 0) ->
+    ok;
+dict_fetch(Dict, NumCycles) ->
+    1 = dict:fetch(1, Dict),
```
These looks like an assertion. Worth to replace with ?assertEqual in order to receive nicer failure reason instead of generic badmatch. Same is true for the other assignment-as-a-match below.

eiri replied:
Those are actually not assertions, but helpers for basic perf test of khash vs dict, i.e. we are not doing functional testing for dict of khash fetch here. But sure, no harm in converting them to the asserts, I'll update that.

kxepal replied:
Well, I agree with you and basically `1 = dict:fetch(1, Dict)` plays good assert role out of the box, so ?assert\* macros aren't required, but they helps much when things goes wrong and imho tests should help determine what and why is broken.

kxepal commented on test/khash_test.erl:
```diff
@@ -0,0 +1,443 @@
+                    {1.0, fun run_put/1},
+                    {1.0, fun run_del/1},
+                    {0.5, fun run_size/1},
+                    % {0.3, fun run_keys/1},
```
What's the reason for having this one commented?

eiri replied:
It was commented in the original etap tests, I kept it for the reference.

kxepal replied:
Hm..ok. Still interesting why.

eiri replied:
I scanned the lib's source and it seems that `khash:keys/1` is planned, but not implemented feature. My guess is that this is sort of a placeholder/reminder.

kxepal commented on test/khash_test.erl:
```diff
@@ -0,0 +1,443 @@
+
+-compile([export_all]).
+
+-define(NODEBUG, true).
```
NODEBUG [implies](https://github.com/erlang/otp/blob/maint/lib/eunit/include/eunit.hrl#L52-L57) NOASSERT which turns all the assertions into pumpkins.

eiri replied:
But only when the tests ran after midnight... Just kidding.

NODEBUG implies NOASSERT only if testing is not enabled and rebar always compiles the tests with `-DTEST`. So NODEBUG here acts as a turn off for debugging macros only.

kxepal replied:
Ah, good to know!

kxepal replied:
Hm, it seems this option prevents tests from run:

```
rebar compile
==> couchdb-khash (compile)
rebar eunit
==> couchdb-khash (eunit)
test/khash_test.erl:26: Warning: variable 'C' is unused
test/khash_test.erl:87: Warning: erlang:now/0: Deprecated BIF. See the "Time and Time Correction in Erlang" chapter of the ERTS User's Guide for more information.
test/khash_test.erl:102: Warning: variable 'State' is unused
test/khash_test.erl:123: Warning: variable 'DTime' is unused
test/khash_test.erl:124: Warning: variable 'KTime' is unused
test/khash_test.erl:133: Warning: variable 'DTime' is unused
test/khash_test.erl:134: Warning: variable 'KTime' is unused
test/khash_test.erl:155: Warning: variable 'H' is unused
test/khash_test.erl:155: Warning: variable 'I' is unused
test/khash_test.erl:168: Warning: variable 'Fold' is unused
test/khash_test.erl:168: Warning: a term is constructed, but never used
test/khash_test.erl:189: Warning: variable 'I' is unused
test/khash_test.erl:189: Warning: variable 'KVs' is unused
test/khash_test.erl:211: Warning: variable 'Err' is unused
test/khash_test.erl:211: Warning: variable 'I' is unused
test/khash_test.erl:218: Warning: variable 'Err' is unused
test/khash_test.erl:218: Warning: variable 'I' is unused
test/khash_test.erl:225: Warning: variable 'Err' is unused
test/khash_test.erl:225: Warning: variable 'I' is unused
test/khash_test.erl:249: Warning: variable 'Err' is unused
test/khash_test.erl:249: Warning: variable 'I' is unused
test/khash_test.erl:256: Warning: variable 'Err' is unused
test/khash_test.erl:256: Warning: variable 'I' is unused
test/khash_test.erl:263: Warning: variable 'Err' is unused
test/khash_test.erl:263: Warning: variable 'I' is unused
test/khash_test.erl:270: Warning: variable 'Err' is unused
test/khash_test.erl:270: Warning: variable 'I' is unused
test/khash_test.erl:277: Warning: variable 'Err' is unused
test/khash_test.erl:368: Warning: variable 'Value' is unused
test/khash_test.erl:380: Warning: variable 'Value' is unused
test/khash_test.erl:384: Warning: variable 'Val' is unused
test/khash_test.erl:413: Warning: variable 'DKeys' is unused
test/khash_test.erl:415: Warning: variable 'HKeys' is unused
test/khash_test.erl:420: Warning: variable 'DKVs' is unused
test/khash_test.erl:421: Warning: variable 'HKVs' is unused
Compiled test/khash_test.erl
======================== EUnit ========================
module 'khash'
module 'khash_test'
  khash_test:16: load_test_ (Loaded khash)...ok
  khash basic operations
    khash_test:30: basic_test_ (Lookup missing is ok)...ok
    khash_test:34: basic_test_ (Get missing is ok)...ok
    khash_test:38: basic_test_ (Del missing is ok)...ok
    khash_test:42: basic_test_ (Stored a key)...ok
    khash_test:46: basic_test_ (Lookuped a key)...ok
    khash_test:50: basic_test_ (Retrieved a key)...ok
    khash_test:54: basic_test_ (Stored a key)...ok
    khash_test:58: basic_test_ (Correct size for hash)...ok
    khash_test:62: basic_test_ (Deleted a key)...ok
    khash_test:66: basic_test_ (Correct size after delete)...ok
    khash_test:70: basic_test_ (Cleared the hash)...ok
    khash_test:74: basic_test_ (Correct size after clear)...ok
    [done in 0.036 s]
  khash randomized test
    khash_test:105: randomized_test_ (State matches dict implementation)...ok
    [done in 0.003 s]
  khash vs dict
    khash_test:122: compare_dict_test_ (Dict's fetch is slower than of khash)...[0.194 s] ok
    khash_test:132: compare_dict_test_ (Dict's store is slower than of khash)...[0.629 s] ok
    [done in 0.833 s]
  khash itrators basics operations
    khash_test:159: basic_iterators_test_ (Got only kv pair as first element)...ok
    khash_test:163: basic_iterators_test_ (Only the one kv pair exists)...ok
    khash_test:167: basic_iterators_test_ (Fold works)...ok
    [done in 0.009 s]
  khash iterators multi-read test
    khash_test:192: multiread_iterators_test_ (Read the same exact key/val pairs)...ok
    [done in 0.003 s]
  khash iterators exipiration functions
    khash_test:215: expiration_iterators_test_ (put expires iterators)...ok
    khash_test:222: expiration_iterators_test_ (del expires iterators)...ok
    khash_test:229: expiration_iterators_test_ (clear expires iterators)...ok
    [done in 0.009 s]
  khash iterators no exipiration functions
    khash_test:253: no_expiration_iterators_test_ (to_list doesn't expire iterators)...ok
    khash_test:260: no_expiration_iterators_test_ (lookup doesn't expire iterators)...ok
    khash_test:267: no_expiration_iterators_test_ (get doesn't expire iterators)...ok
    khash_test:274: no_expiration_iterators_test_ (size doesn't expire iterators)...ok
    khash_test:283: no_expiration_iterators_test_ (iteration doesn't expire iterators)...ok
    [done in 0.015 s]
  [done in 1.916 s]
module 'gen_term'
=======================================================
  All 28 tests passed.
```

If I remove it feels that tests are actually being run:

```
rebar compile
==> couchdb-khash (compile)
rebar eunit
==> couchdb-khash (eunit)
test/khash_test.erl:86: Warning: erlang:now/0: Deprecated BIF. See the "Time and Time Correction in Erlang" chapter of the ERTS User's Guide for more information.
Compiled test/khash_test.erl
======================== EUnit ========================
module 'khash'
module 'khash_test'
  khash_test:15: load_test_ (Loaded khash)...[0.002 s] ok
  khash basic operations
    khash_test:29: basic_test_ (Lookup missing is ok)...ok
    khash_test:33: basic_test_ (Get missing is ok)...ok
    khash_test:37: basic_test_ (Del missing is ok)...ok
    khash_test:41: basic_test_ (Stored a key)...ok
    khash_test:45: basic_test_ (Lookuped a key)...ok
    khash_test:49: basic_test_ (Retrieved a key)...ok
    khash_test:53: basic_test_ (Stored a key)...ok
    khash_test:57: basic_test_ (Correct size for hash)...ok
    khash_test:61: basic_test_ (Deleted a key)...ok
    khash_test:65: basic_test_ (Correct size after delete)...ok
    khash_test:69: basic_test_ (Cleared the hash)...ok
    khash_test:73: basic_test_ (Correct size after clear)...ok
    [done in 0.036 s]
  khash randomized test
    khash_test:104: randomized_test_ (State matches dict implementation)...[2.410 s] ok
    [done in 2.413 s]
  khash vs dict
    khash_test:121: compare_dict_test_ (Dict's fetch is slower than of khash)...test/khash_test.erl:124:<0.75.0>: Dict:     2223065
test/khash_test.erl:125:<0.75.0>: KHash:    1673265
[3.902 s] ok
    khash_test:131: compare_dict_test_ (Dict's store is slower than of khash)...test/khash_test.erl:134:<0.75.0>: Dict:      521395
test/khash_test.erl:135:<0.75.0>: KHash:     144867
[0.666 s] ok
    [done in 4.579 s]
  khash itrators basics operations
    khash_test:158: basic_iterators_test_ (Got only kv pair as first element)...ok
    khash_test:162: basic_iterators_test_ (Only the one kv pair exists)...ok
    khash_test:166: basic_iterators_test_ (Fold works)...ok
    [done in 0.009 s]
  khash iterators multi-read test
    khash_test:191: multiread_iterators_test_ (Read the same exact key/val pairs)...ok
    [done in 0.003 s]
  khash iterators exipiration functions
    khash_test:214: expiration_iterators_test_ (put expires iterators)...ok
    khash_test:221: expiration_iterators_test_ (del expires iterators)...ok
    khash_test:228: expiration_iterators_test_ (clear expires iterators)...ok
    [done in 0.009 s]
  khash iterators no exipiration functions
    khash_test:252: no_expiration_iterators_test_ (to_list doesn't expire iterators)...ok
    khash_test:259: no_expiration_iterators_test_ (lookup doesn't expire iterators)...ok
    khash_test:266: no_expiration_iterators_test_ (get doesn't expire iterators)...ok
    khash_test:273: no_expiration_iterators_test_ (size doesn't expire iterators)...ok
    khash_test:282: no_expiration_iterators_test_ (iteration doesn't expire iterators)...ok
    [done in 0.015 s]
  [done in 8.073 s]
module 'gen_term'
=======================================================
  All 28 tests passed.
```

Thoughts?

eiri replied:
Well, I use rebar 2.5.1 under Ubuntu, tried both to R14B01 and R17.6.5 and I definitely got tests compiled with TEST flag

```
1> ls().
gen_term.beam       gen_term.erl        khash.beam          
khash.erl           khash_test.beam     khash_test.erl      

ok
2> beam_lib:chunks("khash_test.beam", [compile_info]).
{ok,{khash_test,[{compile_info,[{options,[export_all,
                                          {outdir,".eunit"},
                                          {d,'TEST'},
                                          debug_info,debug_info,
                                          {i,"include"},
                                          return_errors,return_warnings,export_all]},
                                {version,"4.7.2"},
                                {time,{2015,7,14,22,29,36}},
                                {source,"/home/vagrant/git/couchdb/src/khash/test/khash_test.erl"}]}]}}
```

My runtime 5 sec which kind of feels like tests are actually running and if I screw with the assertions they are properly failing.

Can you please check compile_info on your test's beam? I mean I don't mind to remote `NODEBUG` flag, I'm not that attached to it, but just for sake of understanding of what's going on here.

kxepal replied:
Ok, I found how I miss this moment early - it's because Erlang 18. Same rebar 2.5.1:

```
{ok,{khash_test,[{compile_info,[{options,[{outdir,".eunit"},
                                          {d,'TEST'},
                                          debug_info,
                                          {i,"include"}]},
                                {version,"6.0"},
                                {time,{2015,7,15,6,8,10}},
                                {source,"/home/kxepal/projects/asf/couchdb/couchdb-khash/test/khash_test.erl"}]}]}}
```

With Erlang R14B01:

```
{ok,{khash_test,[{compile_info,[{options,[export_all,
                                          {outdir,".eunit"},
                                          {d,'TEST'},
                                          debug_info,
                                          {i,"include"},
                                          return_errors,return_warnings,export_all]},
                                {version,"4.7.2"},
                                {time,{2014,10,14,21,1,55}},
                                {source,"/root/couchdb/src/khash/test/khash_test.erl"}]}]}}
```

For the last one tests runs for the same 8 seconds, no warnings. For the Erlang 18 some major amount of tests seems to be skipped. The only major change for eunit was [hrl split](https://github.com/erlang/otp/commit/c574bd33c39d91c487c3fcd819226ecfc46c13c8) but I didn't notice how it could be related.

eiri replied:
I can reproduce the issue with 18.0.2. Apparently R18 compiler ignores `export_all` flag, not sure by design or is it a bug.

I'll remove `NODEBUG` just to simplify things. Thank you for checking this for me.

kxepal replied:
I didn't like this definition from the start ((: You're welcome!

kxepal commented:
I have a strange feeling that most of run_\* and others helper functions are need to convert into proper test functions, but not sure about since made only quick look over the your PR. Will return back tomorrow with clear cache (:

Nice work anyway! DROP ALL ETAP! \o/

eiri committed:
Use assertions in khash's and dict's fetch helpers

eiri committed:
Use assertions in the run helpers of the randomized test

eiri commented:
@kxepal : I've updated fetch and run_\* helpers to use the assertions.

Just a note: originally I implemented randomized test as a lazy generator, but it was disappointedly slow -  30 sec vs 2 sec in a current run. So I know it looks a bit awkward now, but there are a reason for that.

eiri committed:
Remove NODEBUG flag

kxepal commented:
+1

eiri commented:
Merged as per https://github.com/apache/couchdb-khash/commit/6938d72be999c4fb876aa890c1c60105216dc51b

Closing.
